CMakeLists.txt

```diff
@@ -378,167 +437,161 @@ file(
-  ${VSOMEIP_NAME}-e2e PROPERTIES
+  vsomeip3-e2e PROPERTIES
   VERSION ${VSOMEIP_VERSION}
   SOVERSION ${VSOMEIP_MAJOR_VERSION}
+  OUTPUT_NAME ${VSOMEIP_NAME}-e2e
+  EXPORT_NAME vsomeip3-e2e
 )
-if(MSVC)
-  set_target_properties(${VSOMEIP_NAME}-e2e PROPERTIES COMPILE_DEFINITIONS "VSOMEIP_DLL_COMPILATION_PLUGIN")
-endif()
-
+target_compile_definitions(vsomeip3-e2e PUBLIC $<$<CXX_COMPILER_ID:MSVC>:VSOMEIP_DLL_COMPILATION_PLUGIN>)
 target_link_libraries(
-  ${VSOMEIP_NAME}-e2e
-  ${VSOMEIP_NAME}
-  ${Boost_LIBRARIES}
-  ${USE_RT}
-  ${DL_LIBRARY}
-  ${SystemD_LIBRARIES}
+  vsomeip3-e2e
+  PUBLIC vsomeip3
+  PRIVATE RT_INTERFACE DL_INTERFACE OS_INTERFACE
 )
 
-if(${CMAKE_SYSTEM_NAME} MATCHES "QNX")
-  target_link_directories(${VSOMEIP_NAME} PUBLIC "${QNX_TARGET}/${CPUVARDIR}/io-sock/lib")
-  target_link_libraries(${VSOMEIP_NAME}-e2e socket)
-  target_link_libraries(${VSOMEIP_NAME}-cfg socket)
-  target_link_libraries(${VSOMEIP_NAME}-sd socket)
-  target_link_libraries(${VSOMEIP_NAME} PRIVATE socket)
-endif()
 # ######################################################################################################################
 # Compatibility library
 # ######################################################################################################################
-if(VSOMEIP_ENABLE_COMPAT EQUAL 1)
+if(ENABLE_COMPAT)
   set(VSOMEIP_COMPAT_MAJOR_VERSION 2)
   set(VSOMEIP_COMPAT_VERSION ${VSOMEIP_COMPAT_MAJOR_VERSION}.99.99)
 
-  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -DVSOMEIP_ENABLE_COMPAT")
+  add_definitions(-DVSOMEIP_ENABLE_COMPAT)
 
   file(
     GLOB_RECURSE
-    ${VSOMEIP_COMPAT_NAME}_SRC
+    vsomeip3_compat_SRC
     "implementation/compat/logging/src/*.cpp"
     "implementation/compat/message/src/*.cpp"
     "implementation/compat/runtime/src/*.cpp"
   )
-  list(SORT ${VSOMEIP_COMPAT_NAME}_SRC)
+  list(SORT vsomeip3_compat_SRC)
 
-  add_library(${VSOMEIP_COMPAT_NAME} SHARED ${${VSOMEIP_COMPAT_NAME}_SRC})
-  target_compile_features(${VSOMEIP_COMPAT_NAME} PRIVATE cxx_std_17)
+  add_library(vsomeip3-compat SHARED ${vsomeip3_compat_SRC})
+  target_compile_features(vsomeip3-compat PRIVATE cxx_std_17)
+  target_compile_definitions(vsomeip3-compat PUBLIC $<$<CXX_COMPILER_ID:MSVC>:VSOMEIP_DLL_COMPILATION_PLUGIN>)
   set_target_properties(
-    ${VSOMEIP_COMPAT_NAME} PROPERTIES
+    vsomeip3-compat PROPERTIES
     VERSION ${VSOMEIP_COMPAT_VERSION}
     SOVERSION ${VSOMEIP_COMPAT_MAJOR_VERSION}
+    OUTPUT_NAME ${VSOMEIP_COMPAT_NAME}
+    EXPORT_NAME vsomeip3-compat
   )
-  if(MSVC)
-    set_target_properties(${VSOMEIP_COMPAT_NAME} PROPERTIES COMPILE_DEFINITIONS "VSOMEIP_DLL_COMPILATION_PLUGIN")
-  endif()
-
   target_include_directories(
-    ${VSOMEIP_COMPAT_NAME}
+    vsomeip3-compat
     PUBLIC $<BUILD_INTERFACE:${PROJECT_SOURCE_DIR}/interface/compat>
            $<BUILD_INTERFACE:${CMAKE_CURRENT_BINARY_DIR}> # for generated files in build mode
+           $<BUILD_INTERFACE:${PROJECT_BINARY_DIR}/implementation/configuration/include>
```

Comment:
Yes!

So, in the current version:
```cmake
configure_file (
   "${PROJECT_SOURCE_DIR}/implementation/configuration/include/internal.hpp.in"
   "${PROJECT_SOURCE_DIR}/implementation/configuration/include/internal.hpp"
)
```
This writes `internal.hpp` into the source directory.  I would argue that this is a build artifact though and should go into the build directory.

Especially because it's in the `.gitignore`, when there are changes to this file they're hard to see.  When it's in the build directory it's trivial to delete the build directory and get a fresh build.  When it's in the source directory though it persists.  CMake should keep it updated, but I'm pretty sure what first brought my attention to this was the cases where it wasn't updated.

So the change here is to put it in the build directory:
```cmake
configure_file(
  "${PROJECT_SOURCE_DIR}/implementation/configuration/include/internal.hpp.in"
  "${PROJECT_BINARY_DIR}/implementation/configuration/include/internal.hpp"
)
```
(key being `${PROJECT_BINARY_DIR}`)

And then the line you pointed out adds that path in the build directory to the include path.

The reason it causes me so much trouble is I have it in the binary directory in my forks, but when I switch back and forth from my forks to the upstream code that file re-appears in the source directory, and then overrides the version in the build directory.

By having it in the build directory, my builds seem more robust.  Especially when I'm testing certain compiler flags, and have to call `clang++` or `q++` directly (without CMake) to see the different effects.  Having `internal.hpp` be in the build directory instead of the source makes this a lot easier.